proxy/contracts/schain/MessageProxyForSchain.sol

```diff
@@ -203,22 +217,39 @@ contract MessageProxyForSchain is MessageProxy, IMessageProxyForSchain {
         for (uint256 i = 0; i < messages.length; i++) {
             _callReceiverContract(fromChainHash, messages[i], startingCounter + 1);
         }
-        _topUpBalance();
+        _topUpSenderBalance();
     }
 
     /**
      * @dev Sets new version of contracts on schain
-     * 
+     *
      * Requirements:
-     * 
+     *
      * - `msg.sender` must be granted DEFAULT_ADMIN_ROLE.
      */
-    function setVersion(string calldata newVersion) external override {
-        require(hasRole(DEFAULT_ADMIN_ROLE, msg.sender), "DEFAULT_ADMIN_ROLE is required");
+    function setVersion(string calldata newVersion) external override onlyOwner {
         emit VersionUpdated(version, newVersion);
         version = newVersion;
     }
 
+    function setMinimumReceiverBalance(uint256 balance) external override onlyConstantSetter {
+        minimumReceiverBalance = balance;
+    }
+
+    function topUpReceiverBalance(address payable receiver) external override {
+        // allow only TokenManager to call this function
+        require(getTokenManagerLinker().hasTokenManager(msg.sender), "Sender is not TokenManager");
```

Comment:
I would suggest to check registered contracts for all instead of token manager only.
First of all it is a bigger or equal set of contracts than set of TokenManagers.
Set of registered contracts for all is a set of contracts which is possible to call from every chain including Mainnet.
Second it will produce less amount of code to use registered contract for all.

Or maybe even check if this contract registered for a current sender chain.
I think we need to include @cstrangedk here

Reply:
Refactored.